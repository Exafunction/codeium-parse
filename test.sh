#!/bin/bash
set -euo pipefail

cd "$(dirname "${BASH_SOURCE[0]}")"
for test_file in test_files/*; do
    test_file="$(basename "$test_file")"
    echo "$test_file"
    ./parse -file "test_files/$test_file" -use_tags_query -tags_query_dir "queries" > "goldens/$test_file.golden.tmp"
    trap 'rm "goldens/$test_file.golden.tmp"' EXIT
    diff -u "goldens/$test_file.golden" "goldens/$test_file.golden.tmp"
done
