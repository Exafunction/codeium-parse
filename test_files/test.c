#include "runtime/parser.h"

typedef struct {
  Parser* parser;
  TSSymbol lookahead_symbol;
} SkipPrecedingTreesSession;

static void parser__push(Parser* self, StackVersion version, Tree* tree,
                         TSStateId state) {
  ts_stack_push(self->stack, version, tree, false, state);
  ts_tree_release(tree);
}

static Tree* parser__lex(Parser* self, StackVersion version) {
  TSStateId parse_state = ts_stack_top_state(self->stack, version);
  Length start_position = ts_stack_top_position(self->stack, version);
  TSLexMode lex_mode = self->language->lex_modes[parse_state];
}
