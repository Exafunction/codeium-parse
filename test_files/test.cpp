#include <string>

namespace bar {
// Declaration.
bool baz(const std::string& bing);

struct Foo2;

}  // namespace bar

namespace bar::bar2 {
struct Foo3;
}

class Foo {
 public:
  // foo
  // foo2
  void foo() {
    // Do nothing.
  }
};

struct bar::Foo2 {};
struct bar::bar2::Foo3 {};

// baz
bool bar::baz(const std::string& bing) {
  // This shouldn't be recognized as a function.
  Foo dist();
  return bing.empty();
}

// one_pointer comment
bool* one_pointer() { return nullptr; }
// two_pointer comment
bool** two_pointer() { return nullptr; }

// Comment.
template <typename T>
class TemplatedFoo {
  template <typename U>
  void foo();
};

// Comment.
template <typename T>
template <typename U>
void TemplatedFoo<T>::foo() {}

// Comment.
template <>
template <>
void TemplatedFoo<int>::foo<int>() {}

// Comment.
template <typename T>
class TemplatedFoo2;

// Comment.
template <>
class TemplatedFoo2<int> {};
