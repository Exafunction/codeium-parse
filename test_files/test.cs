// clang-format off
namespace Namespace {
  // Comment A.
  internal delegate void A(params int[] test);
  // Comment F.
  public struct F<T>
      where T : struct {
  }
  // Comment F.
  [Nice]
  private record F<T1, T2>
      where T1 : I1, I2, new()
      where T2 : I2 {}
  // Comment Teacher.
  record Teacher(string FirstName, string LastName, string Subject) : Person(FirstName, LastName);
  // Comment B.
  enum B { Ten = 10, Twenty = 20 }
  // Comment F.
  public class F : object, IAlpha, IOmega {}
  // Comment Class.
  public partial class Class<in TParam>
      where TParam : class ?, notnull, F ? {
    // Comment operator +.
    public static int operator +(A a) {
      return 0;
    }
    // Comment GetSet.
    uint GetSet { get; set; }
    // Comment Foo.
    static extern Foo() => bar = 0;
    // Comment ~Class.
    extern ~Class() {}
    // Comment Bar
    public void Bar() => bar = 0;
  }
  // Comment IFoo.
  public interface IFoo {}
}
// clang-format on
