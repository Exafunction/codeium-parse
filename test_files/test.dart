import 'package:test/test.dart';

// Functions

// foo not a docstring
void foo() {
    print("Hello, World!");
}

/// bar is a docstring
int bar() {
    return 0;
}

// Classes with methods

class FooClass {
    int someField = 0;

    // Constructor
    FooClass() {
        this.someField = 0;
    }

    /// FCB docstring
    int fooClassBar() {
        return 0;
    }
}

// Mixins
mixin FooMixin {
    int barMixin() {
        return 0;
    }
}

// Combinations
class FooBar extends FooClass with Foo {
    int bar() {
        return 0;
    }
}

void main() {
    print("Hello, World!");
}
