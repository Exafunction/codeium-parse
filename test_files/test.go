package test_files

// Pattern 0.
type Pattern0 struct {
	a string
}

// Pattern 1.
type Pattern1 interface {
	a() string
}

// Unsupported.
type (
	UnsupportedStruct    struct{}
	UnsupportedInterface interface{}
)

func Pattern2() {}

// Comments
// here.
func (p *Pattern0) Method(string) {}

func (p Pattern0) Method2(a string) string {
	Pattern2()
	s := &Pattern0{}
	s.Method("")
	((*Pattern0)(s)).Method("")
	var s2 any
	s2.(*Pattern0).Method("")
	return a
}

type Alias string
