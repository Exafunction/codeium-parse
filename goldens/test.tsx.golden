Name: "path/to/file/1"
Definition (definition.import):
import { named1, named2 as alias2, named3 as alias3 } from "path/to/file/1";
Named imports:
    named1
    named2 as alias2
    named3 as alias3

Name: "path/to/file/2"
Definition (definition.import):
import default1 from "path/to/file/2";
Default import: default1

Name: "path/to/file/3"
Definition (definition.import):
import { named4 as alias4, named5 as alias5, type NamedType1 as AliasType1 } from "path/to/file/3";
Named imports:
    named4 as alias4
    named5 as alias5
    NamedType1 as AliasType1

Name: 'npm-lib-1'
Definition (definition.import):
import default2, { named4 as alias4 } from 'npm-lib-1'
Default import: default2
Named imports:
    named4 as alias4

Name: "npm-lib-2"
Definition (definition.import):
import * as entireModule from "npm-lib-2";
Namespace: entireModule

Name: "path/to/file/4"
Definition (definition.import):
import type { NamedType2 } from "path/to/file/4";
Named imports:
    NamedType2

Name: 'npm-lib-3'
Definition (definition.import):
import default3 = require('npm-lib-3');
Default import: default3

Name: "./side-effects-only-1.ts"
Definition (definition.import):
import "./side-effects-only-1.ts";

Name: f
Parameters: (a: string)
Return type: string
Doc:
/** f comment. */
Declaration (definition.function):
function f(a: string): string;

Name: f
Parameters: (a: any)
Return type: any
Definition (definition.function):
function f(a: any): any {
    import default4 from 'npm-lib-4';
    return a;
}

Name: 'npm-lib-4'
Definition (definition.import):
import default4 from 'npm-lib-4';
Lineage: [f]
Lineage types: [function]
Default import: default4

Name: f2
Parameters: (a: number)
Return type: void
Doc:
/** f2 comment. */
Declaration (definition.function):
declare function f2(a: number): void;

Name: f3
Parameters: ()
Return type: void
Doc:
/** f3 comment. */
Declaration (definition.function):
export function f3(): void;
Metadata:
    is_export = true

Name: f3
Parameters: ()
Return type: void
Definition (definition.function):
export function f3(): void {}
Metadata:
    is_export = true

Name: f4
Parameters: ()
Return type: void
Doc:
/** f4 comment. */
Declaration (definition.function):
export declare function f4(): void;
Metadata:
    is_export = true

Name: Foo
Doc:
/** Foo comment. */
Definition (definition.class):
class Foo {
    /** methodSignature comment. */
    methodSignature(): void;
    /** methodSignature comment. */
    methodSignature(): void {}
}

Name: methodSignature
Parameters: ()
Return type: void
Doc:
/** methodSignature comment. */
Declaration (definition.method):
methodSignature(): void
Lineage: [Foo]
Lineage types: [class]

Name: methodSignature
Parameters: ()
Return type: void
Doc:
/** methodSignature comment. */
Definition (definition.method):
methodSignature(): void {}
Lineage: [Foo]
Lineage types: [class]

Name: Foo2
Doc:
/** Foo2 comment. */
Definition (definition.class):
declare class Foo2 {
    /** methodSignature2 comment. */
    methodSignature2(): void;
}

Name: methodSignature2
Parameters: ()
Return type: void
Doc:
/** methodSignature2 comment. */
Declaration (definition.method):
methodSignature2(): void
Lineage: [Foo2]
Lineage types: [class]

Name: Foo3
Doc:
// *Foo3 comment.""
Definition (definition.class):
export declare class Foo3 {}
Metadata:
    is_export = true

Name: Module
Doc:
/** Module comment. */
Definition (definition.module):
module Module {}

Name: Module2
Doc:
/** Module2 comment. */
Definition (definition.module):
export module Module2 {}
Metadata:
    is_export = true

Name: 'module'
Doc:
/** Module3 comment. */
Definition (definition.module):
declare module 'module'

Name: AbstractClass
Doc:
/** AbstractClass comment. */
Definition (definition.class):
abstract class AbstractClass {
    /** Abstract method comment. */
    abstract method(a: string): void;
}

Name: method
Parameters: (a: string)
Return type: void
Doc:
/** Abstract method comment. */
Declaration (definition.method):
abstract method(a: string): void
Lineage: [AbstractClass]
Lineage types: [class]

Name: AbstractClass2
Doc:
/** AbstractClass2 comment. */
Definition (definition.class):
export abstract class AbstractClass2 {}
Metadata:
    is_export = true

Name: AbstractClass3
Doc:
/** AbstractClass3 comment. */
Definition (definition.class):
declare abstract class AbstractClass3 {}

Name: AbstractClass4
Doc:
/** AbstractClass4 comment. */
Definition (definition.class):
export declare abstract class AbstractClass4 {}
Metadata:
    is_export = true

Name: Interface
Doc:
/** Interface comment. */
Definition (definition.interface):
interface Interface {
    /** Interface method comment. */
    interfaceMethod(a: string): void;
}

Name: interfaceMethod
Parameters: (a: string)
Return type: void
Doc:
/** Interface method comment. */
Declaration (definition.method):
interfaceMethod(a: string): void
Lineage: [Interface]
Lineage types: [interface]

Name: Interface2
Doc:
/** Interface2 comment. */
Definition (definition.interface):
export interface Interface2 {}
Metadata:
    is_export = true

Name: Interface3
Doc:
/** Interface3 comment. */
Definition (definition.interface):
declare interface Interface3 {}

Name: Interface4
Doc:
/** Interface4 comment. */
Definition (definition.interface):
export declare interface Interface4 {}
Metadata:
    is_export = true

Name: FooString
Doc:
/** FooString comment. */
Definition (definition.type):
type FooString = string;

Name: FooString2
Doc:
/** FooString2 comment. */
Definition (definition.type):
export type FooString2 = string;
Metadata:
    is_export = true

Name: FooString3
Doc:
/** FooString3 comment. */
Definition (definition.type):
declare type FooString3 = string;

Name: FooString4
Doc:
/** FooString4 comment. */
Definition (definition.type):
export declare type FooString4 = string;
Metadata:
    is_export = true

Name: NamespaceFoo
Definition (definition.namespace):
namespace NamespaceFoo {}

Name: NamespaceFoo2
Doc:
/** NamespaceFoo2 comment. */
Definition (definition.namespace):
export namespace NamespaceFoo2 {}
Metadata:
    is_export = true

Name: NamespaceFoo3
Doc:
/** NamespaceFoo3 comment. */
Definition (definition.namespace):
declare namespace NamespaceFoo3 {}

Name: NamespaceFoo4
Doc:
/** NamespaceFoo4 comment. */
Definition (definition.namespace):
export declare namespace NamespaceFoo4 {}
Metadata:
    is_export = true
